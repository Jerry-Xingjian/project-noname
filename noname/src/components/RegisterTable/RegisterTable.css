.background-white {
    background-color: #fff;
}