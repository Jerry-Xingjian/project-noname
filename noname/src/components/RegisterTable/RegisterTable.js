import React, { useState, useEffect } from 'react';
import './RegisterTable.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import { Link, useNavigate } from 'react-router-dom';
import { register } from '../../utils/api/user';

function RegisterTable(props) {
  const { showAlert } = props;
  const [passwordEqual, setPasswordEqual] = useState(false);
  const [password, setPassword] = useState('');
  const navigate = useNavigate();
  const [confirmedPassword, setConfirmedPassword] = useState('');
  const checkPasswordEqual = () => setPasswordEqual(password === confirmedPassword);

  // Password changes handler
  const handlePasswordChange = (event) => {
    setPassword(event.target.value);
  };

  const handleConfirmPasswordChange = (event) => {
    setConfirmedPassword(event.target.value);
  };

  // Check if password and confirmed password are equal
  useEffect(() => {
    checkPasswordEqual();
  }, [password, confirmedPassword]);

  const handleSubmit = async (event) => {
    event.preventDefault();
    const email = document.getElementById('floatingInput').value;
    try {
      await register(email, password);
      navigate('/home');
    } catch (err) {
      if (err.message.includes(410)) {
        showAlert({ show: true, message: 'Duplicate email. This email has already been registered.' });
      } else {
        showAlert({ show: true, message: 'Connection refused. Check your internet connection.' });
      }
    }
  };

  return (
    <div className="container d-flex flex-column w-25 position-absolute top-50 end-0 translate-middle">
      <form className="d-flex flex-column" onSubmit={handleSubmit}>
        <div className="mb-3">
          <h1 className="text-center">Sign In</h1>
          <p className="text-center text-black-50">Start your journey from here</p>
        </div>
        <div className="form-floating mb-3">
          <input type="email" className="form-control background-white" id="floatingInput" placeholder="name@example.com" name="email" />
          <label htmlFor="floatingInput">Email address</label>
        </div>
        <div className="form-floating mb-3">
          <input type="password" className="form-control background-white" id="floatingPassword" placeholder="Password" name="password" onChange={handlePasswordChange} value={password} />
          <label htmlFor="floatingPassword">Password</label>
        </div>
        <div className="form-floating mb-3">
          <input type="password" className={`form-control background-white ${passwordEqual ? '' : 'is-invalid'}`} id="floatingConfirmedPassword" placeholder="Confirm Password" onChange={handleConfirmPasswordChange} value={confirmedPassword} />
          <label htmlFor="floatingPassword">Confirm your password</label>
        </div>
        <button type="submit" className="btn btn-danger text-light mb-3">Sign In</button>
      </form>
      {/* <button type="button" className="btn btn-primary text-light mb-3">
      Sign in with Google</button> */}
      <div className="text-center text-black-50 d-flex justify-content-center">
        <p className="my-0 me-1">Already have an account?</p>
        <Link className="text-danger" to="/login">
          <p>Login here!</p>
        </Link>
      </div>
    </div>
  );
}

export default RegisterTable;
