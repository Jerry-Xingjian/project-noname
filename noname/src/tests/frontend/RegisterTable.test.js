import { render, screen, fireEvent, waitFor} from "@testing-library/react";
import userEvent from '@testing-library/user-event';
import {createMemoryHistory} from 'history';
import React from "react";
import {Router} from 'react-router-dom'
import '@testing-library/jest-dom'
import RegisterTable from '../../components/RegisterTable/RegisterTable';
import renderer from 'react-test-renderer';
import axios from "axios";
import MockAdapter from "axios-mock-adapter";

describe("Component test", () => {
  test('certain texts exists in the page', () => {
    const history = createMemoryHistory();
    render(
      <Router location={history.location} navigator={history}>
        <RegisterTable />
      </Router>,
    )
    // check whether "Start your journey from here" exits in the screen or not
    const text = screen.getByText(/Start your journey from here/i)
    expect(text).toBeInTheDocument()
  })
});

describe("Snapshot test", () => {
  test('RegisterTable component matches snapshot', () => {
    const history = createMemoryHistory();
    const component = renderer.create(
      <Router location={history.location} navigator={history}>
        <RegisterTable />
      </Router>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Navigation test", () => {
  let mock;

  beforeAll(() => {
    mock = new MockAdapter(axios);
  });

  afterEach(() => {
    mock.reset();
  });

  const default_data = [{ test_data: 20}];

  test("Register navigate to login page", async () => {
    // test for different effects than the page actually changing
    const history = createMemoryHistory();
    const spy = jest.spyOn(history, 'push');
    render(
      <Router location={history.location} navigator={history}>
        <RegisterTable />
      </Router>,
    )
    const user = userEvent.setup()
    expect(screen.getByText(/Login here!/i)).toBeInTheDocument()
    await user.click(screen.getByText(/Login here!/i))

    // check that the content changed to the new page
    expect(history.location.pathname).toBe('/login');
    expect(spy).toBeCalled();
  });

  test("Login navigate to home page", async () => {
    // test for different effects than the page actually changing
    const history = createMemoryHistory();
    // spying on history.push
    const spy = jest.spyOn(history, 'push');
    render(
      <Router location={history.location} navigator={history}>
        <RegisterTable />
      </Router>
    )

    // Mock POST request when param are `default_data` and user input email and password
    mock.onPost().reply(200, default_data)
    // waitFor may run the callback function, in this case, fireEvent.click
    // a number of times until the timeout is reached
    // keep calling axios until got response
    await waitFor(() => {
      fireEvent.click(screen.getByRole('button', { name: 'Sign In' }));
    });
    // test whether the axios test is post request or not
    expect(mock.onPost().reply(200, default_data).history.post[0].method).toEqual("post")
    // check that the content changed to the new page
    // expect(spy).toBeCalled();
    expect(history.location.pathname).toBe('/');
  });
});

describe("Event test", () => {
  test("Email, password, confirm password type test", async () => {
    const history = createMemoryHistory();
    render(
      <Router location={history.location} navigator={history}>
        <RegisterTable />
      </Router>
    )
    const email = screen.getByLabelText("Email address");
    const password = screen.getByLabelText("Password");
    const passwordConfirm = screen.getByLabelText("Confirm your password");
    await userEvent.type(email, 'xwang@gmail.com');
    expect(email).toHaveValue('xwang@gmail.com');
    await userEvent.type(password, '123456');
    expect(password).toHaveValue('123456');
    userEvent.type(passwordConfirm, '123456');
    expect(passwordConfirm).toHaveValue('123456');
  });
});


