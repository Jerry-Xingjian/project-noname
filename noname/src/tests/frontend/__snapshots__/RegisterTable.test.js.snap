// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test RegisterTable component matches snapshot 1`] = `
<div
  className="container d-flex flex-column w-25 position-absolute top-50 end-0 translate-middle"
>
  <form
    className="d-flex flex-column"
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <h1
        className="text-center"
      >
        Sign In
      </h1>
      <p
        className="text-center text-black-50"
      >
        Start your journey from here
      </p>
    </div>
    <div
      className="form-floating mb-3"
    >
      <input
        className="form-control background-white"
        id="floatingInput"
        name="email"
        placeholder="name@example.com"
        type="email"
      />
      <label
        htmlFor="floatingInput"
      >
        Email address
      </label>
    </div>
    <div
      className="form-floating mb-3"
    >
      <input
        className="form-control background-white"
        id="floatingPassword"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <label
        htmlFor="floatingPassword"
      >
        Password
      </label>
    </div>
    <div
      className="form-floating mb-3"
    >
      <input
        className="form-control background-white is-invalid"
        id="floatingConfirmedPassword"
        onChange={[Function]}
        placeholder="Confirm Password"
        type="password"
        value=""
      />
      <label
        htmlFor="floatingPassword"
      >
        Confirm your password
      </label>
    </div>
    <button
      className="btn btn-danger text-light mb-3"
      type="submit"
    >
      Sign In
    </button>
  </form>
  <button
    className="btn btn-primary text-light mb-3"
    type="button"
  >
    Sign in with Google
  </button>
  <div
    className="text-center text-black-50 d-flex justify-content-center"
  >
    <p
      className="my-0 me-1"
    >
      Already have an account?
    </p>
    <a
      className="text-danger"
      href="/login"
      onClick={[Function]}
    >
      <p>
        Login here!
      </p>
    </a>
  </div>
</div>
`;
